src/core/view.cc

```diff
@@ -237,20 +201,41 @@ View::set_not_visible(Download* download) {
 }
 
 void
-View::next_focus() {
+View::next_focus(unsigned int i) {
   if (empty())
     return;
 
-  m_focus = (m_focus + 1) % (size() + 1);
+  // If at the boundary, roll over
+  if (m_focus == size() - 1) {
+    m_focus = size();
+  } else {
+    // Move forward, stop at the boundary
+    if (m_focus == size())
+      m_focus = i - 1;
+    else
+      m_focus += i;
+    if (m_focus > size() - 1)
+      m_focus = size() - 1;
+  }
```

Comment:
This looks wrong and is messy.

```
if (empty() || i == 0)
  return;

if (m_focus == size() - 1) {
  m_focus = 0;
  emit_changed();
  return;
}

m_focus += i;

if (m_focus > size() - 1)
  m_focus = size() - 1;

emit_changed();
```

Same for prev_focus.

Reply:
Maybe I misunderstood, but this is the logic required to support the following UX for a page down of 50 with 146 items:
`none (default state) -> 50 -> 100 -> 146 ("next" boundary) -> none`
For page up:
`none -> 146 -> 96 -> 46 -> none ("prev" boundary)`
while keeping KEY_DOWN/KEY_UP the same. I did try that code early on and found that it will never be able to reach m_focus == size(), which makes KEY_DOWN work differently than it currently does. If you want to change the flow so that page up/down are the ones who can never reach size(), I can split things out into two functions.

I have fixed up the style to match, though, and moved ^K to ^U to fix a key binding conflict.

Reply:
Should probably have both be able to reach `size()`, kind of forgot how it worked.